### ISSUE TITLE
"Error:  unavailable" in  runner.js:29:9 when using dojo/domReady!

### ISSUE THREAD

**Issue by vaxul** [OpenNTF/dojo-webpack-plugin#257]
As soon I use `dojo/domReady!` within the define declaration I get an the following error:
```
Uncaught Error:  unavailable
    at runner.js:29
    at loaderProxy:2
    at Object.<anonymous> (loaderProxy:3)
    at u (bootstrap:19)
    at Object.<anonymous> (test.js:2)
    at u (bootstrap:19)
    at Object.<anonymous> (bootstrap.js:2)
    at u (bootstrap:19)
    at bootstrap:324
    at bundle.1570462823.js:1
```

boostrap.js:
```js
define([
  'ui/test',
], function(
    Test
) {
  console.log('Test', Test);
});
```

ui/test.js
```js
define([
  'dojo/domReady!',
], function(
) {
  return {
    property: 'Test',
  };
});
```
I tested this with the dojo-webpack-plugin-sample-master.

Before I could run the webpack build with dojo/domReady, I had to add the snippet to the webpackconfig as @chuckdumont mentioned in https://github.com/OpenNTF/dojo-webpack-plugin/issues/167#issuecomment-388062655

Maybe I have to follow @chuckdumont advice from https://github.com/OpenNTF/dojo-webpack-plugin/issues/167#issuecomment-388330508:
> Also, you should remove js-src from the resolveLoader modules so that webpack won't try to load any Dojo loader plugins as webpack loaders.

But I not sure what this means.

**Comment by chuckdumont** [comment]
Hmm.  I think that comment was intended for a different issue.  I removed it.  `resolveLoader` should not be used to point to dojo loaders.  It is for webpack loaders only.

I also added an updated to #167 about the [async](https://github.com/OpenNTF/dojo-webpack-plugin#async) config property and how it can be used to eliminate the mentioned requirement that the DOM be already loaded when using dojo/loaderProxy with dojo/domReady!

**Comment by vaxul** [comment]
Hi, thank you for the update.

Adding the async option resolves the dojo/domReady! problem:
```
{
	// ...
	plugins: [
		new DojoWebpackPlugin({
			// ...
			async: true,
		}),

		// ...

		// CUSTOM
		new webpack.NormalModuleReplacementPlugin(
				/^dojo\/domReady!/, (data) => {
					const match = /^dojo\/domReady!(.*)$/.exec(data.request);
					data.request = "dojo/loaderProxy?loader=dojo/domReady!" + match[1];
				}
		),
	],
	// ...
}
```

But this introduce another problem to legacy code. Each module, which requires `dojo/domReady!` must be wrapped by `Promise.resolve`.

Is there any possibility to resolve modules with dojo/domReady without manually touch the legacy code?

**Comment by chuckdumont** [comment]
Unless there's a bug that I'm not aware of, you don't need to use `Promise.resolve` in legacy (i.e. Dojo) code.  You only need to use it when using `require` in CommonJS code (which is a synchronous function) to load AMD modules.  There isn't any way around this.  You can't use a synchronous API to load modules asynchronously without promises.   If that doesn't work for you, then the only other options are to disable `async` mode and:

* Replace `dojo/domReady` with `dojo/ready`.  `dojo/ready` is not a loader plugin so no issues with module resolution.
* Ensure that the DOM is ready before loading any modules that use `dojo/domReady`.  That way, `dojo/domReady` will resolve synchronously.

**Comment by vaxul** [comment]
Ok - to summarize:

It's necessary to wrap `Promise.resolve` around legacy require statements from the legacy modules.

Before like:
```
if (something) {
	require('moduleWhichUsingDojoDomReady', function(moduleWhichUsingDojoDomReady) {
		moduleWhichUsingDojoDomReady.init();
	});
}
```

Must do after:
```
if (something) {
	Promise.resolve(require('moduleWhichUsingDojoDomReady')).then(function(moduleWhichUsingDojoDomReady) {
		moduleWhichUsingDojoDomReady.init();
	});
}
```

Or relace `dojo/domReady` with `dojo/ready`. <- I will stick to this.

Thank you for your help. 👍

**Comment by chuckdumont** [comment]
No.  You don't need to do anything to legacy (Dojo) modules.  You only need to use `Promise.resolve` in **CommonJS** modules, if you are using `require` to load an AMD module and `async` mode is enabled.

**Comment by chuckdumont** [comment]
Your example 'before' code is wrong.  You seem to be confusing AMD's async `require` with CommonJS `require` and Dojo's synchronous `require`.

You're specifying a callback function as the second argument, implying async `require` but the first argument is not a string array, so it won't be treated as async `require`.  Before you go on, I suggest you become familiar with the differences between AMD and CommonJS module formats and how `require` is used in each.  Also, note how dojo-webpack-plugin handles [CommonJS require vs. Dojo's synchronous require](https://github.com/OpenNTF/dojo-webpack-plugin#commonjs-require-vs-dojo-synchronous-require) which both have the same signature.

As I said in the previous comment, you shouldn't need to make any changes in your existing AMD modules in order to use `async` mode.
